src/libtrx/memory.c

```diff
@@ -51,3 +52,52 @@ char *Memory_DupStr(const char *const string)
     strcpy(memory, string);
     return memory;
 }
+
+void *Memory_ArenaAlloc(
+    MEMORY_ARENA_ALLOCATOR *const allocator, const size_t size)
+{
+    if (allocator->default_chunk_size == 0) {
+        allocator->default_chunk_size = 1024 * 4; // default to 4K
+    }
+
+    if (allocator->current_chunk == nullptr
+        || allocator->current_chunk->offset + size
+            > allocator->current_chunk->size) {
+        const size_t new_chunk_size = MAX(allocator->default_chunk_size, size);
+        MEMORY_ARENA_CHUNK *const new_chunk =
+            Memory_Alloc(sizeof(MEMORY_ARENA_CHUNK) + new_chunk_size);
+        new_chunk->memory = (char *)new_chunk + sizeof(MEMORY_ARENA_CHUNK);
+        new_chunk->size = new_chunk_size;
+        new_chunk->offset = 0;
+        new_chunk->next = nullptr;
+        if (allocator->current_chunk != nullptr) {
+            allocator->current_chunk->next = new_chunk;
+        }
+        allocator->current_chunk = new_chunk;
+    }
+
+    void *const result = (char *)allocator->current_chunk->memory
+        + allocator->current_chunk->offset;
+    allocator->current_chunk->offset += size;
+    return result;
+}
+
+void Memory_ArenaReset(MEMORY_ARENA_ALLOCATOR *const allocator)
+{
+    MEMORY_ARENA_CHUNK *chunk = allocator->first_chunk;
+    while (chunk != nullptr) {
+        chunk->offset = 0;
+        chunk = chunk->next;
+    }
+    allocator->current_chunk = allocator->first_chunk;
```

Comment:
I think there might be an issue with this code but I'm not sure. From a quick look:

It looks like `Memory_ArenaReset` will set the offset of all chunks to 0 and keep the chain, but `Memory_ArenaAlloc` reaches the end of the current alloc buffer and makes a new one every time. It won't check that a "next" already exists, and will make a new chunk and then set that as the new next, breaking the chain and causing a memory leak.

It should also ensure that the amount requested is in fact smaller than the new chunk, otherwise it will make something too small and then whatever wanted it will trash memory as it goes over the "chunk" size.

Reply:
@walkawayy updated.